auditbeat/module/file_integrity/fileinfo_windows.go

```diff
@@ -65,24 +67,27 @@ func NewMetadata(path string, info os.FileInfo) (*Metadata, error) {
 	// FILE_FLAG_BACKUP_SEMANTICS flag.
 	var err error
 	if !info.IsDir() {
-		fileInfo.SID, fileInfo.Owner, err = fileOwner(path)
+		if fileInfo.SID, fileInfo.Owner, err = fileOwner(path); err != nil {
+			errs = append(errs, errors.Wrap(err, "fileOwner failed"))
+		}
+
 	}
-	fileInfo.Origin, err = GetFileOrigin(path)
-	return fileInfo, err
+	if fileInfo.Origin, err = GetFileOrigin(path); err != nil {
+		errs = append(errs, errors.Wrap(err, "GetFileOrigin failed"))
+	}
+	return fileInfo, errs.Err()
 }
 
 // fileOwner returns the SID and name (domain\user) of the file's owner.
 func fileOwner(path string) (sid, owner string, err error) {
-	f, err := file.ReadOpen(path)
-	if err != nil {
-		return "", "", errors.Wrap(err, "failed to open file to get owner")
-	}
-	defer f.Close()
-
 	var securityID *syscall.SID
 	var securityDescriptor *SecurityDescriptor
 
-	if err = GetSecurityInfo(syscall.Handle(f.Fd()), FileObject,
+	pathW, err := syscall.UTF16PtrFromString(path)
+	if err != nil {
+		return sid, owner, errors.Wrapf(err, "failed to convert path:'%s' to UTF16", path)
+	}
+	if err = GetNamedSecurityInfo(pathW, FileObject,
```

Comment:
The problem here was:
- Some applications will open files for writing in exclusive mode (non FILE_READ_SHARED).
- ReadOpen above wants read access, so it fails with permission denied.

As GetSecurityInfo doesn't need read access to the file, but only the metadata (READ_CONTROL), I tried opening the file with just that flag. This also fails because the underlying CreateFile will internally add more permissions to the request (SYNCHRONIZE|FILE_READ_ATTRIBUTES) which causes it to still fail for files open with exclusivity.

Finally I came across this SO thread: https://stackoverflow.com/questions/58393139/createfilew-for-read-control-fails-with-access-is-denied-despite-being-owner-o that points to using GetNamedSecurityInfo, which internally does an NtOpenFile call with the minimum required permissions.